robot/impl/robot_reconfigure_weak_dependencies_test.go

```diff
@@ -285,4 +291,363 @@ func TestUpdateWeakDependents(t *testing.T) {
 	test.That(t, weak1.resources, test.ShouldHaveLength, 2)
 	test.That(t, weak1.resources, test.ShouldContainKey, base1Name)
 	test.That(t, weak1.resources, test.ShouldContainKey, base2Name)
+	test.That(t, weak1.reconfigCount, test.ShouldEqual, 3)
+}
+
+func TestWeakDependentsExplicitDependency(t *testing.T) {
+	// This test tests the behavior of weak dependents that have an explicit dependency.
+	// The robot has 3 components:
+	//   * base1 and base2: two fake base components.
+	//   * weak1: a weak component that depends on all components and also has an explicit dependency on base1.
+	logger := logging.NewTestLogger(t)
+
+	// Register a `Resource` that generates weak dependencies. Specifically instance of
+	// this resource will depend on every `component` resource.
+	weakAPI := resource.NewAPI(uuid.NewString(), "component", "weak")
+	weakModel := resource.DefaultModelFamily.WithModel(utils.RandomAlphaString(5))
+	weak1Name := resource.NewName(weakAPI, "weak1")
+	resource.Register(
+		weakAPI,
+		weakModel,
+		resource.Registration[*someTypeWithWeakAndStrongDeps, *someTypeWithWeakAndStrongDepsConfig]{
+			Constructor: func(
+				ctx context.Context,
+				deps resource.Dependencies,
+				conf resource.Config,
+				logger logging.Logger,
+			) (*someTypeWithWeakAndStrongDeps, error) {
+				return &someTypeWithWeakAndStrongDeps{
+					Named:     conf.ResourceName().AsNamed(),
+					resources: deps,
+				}, nil
+			},
+			WeakDependencies: []resource.Matcher{resource.TypeMatcher{Type: resource.APITypeComponentName}},
+		})
+	defer resource.Deregister(weakAPI, weakModel)
+
+	// Start robot with two components and one resource with weak dependencies
+	// (that also has an explicit dependency on one of the components).
+	// The following scenario is expected:
+	// 1) The two bases will configure first in parallel (which will each bump the logical clock).
+	// 2) The completeConfig loop will check to see if there is a need to updateWeakDependents by
+	//    checking the logical clock on also whether there are dependencies on weak dependents.
+	//    The check will fail.
```

Comment:
> I think on -> and?

Agree, it should be `and`.

> Which conditions fail?

Failure (no update of weak dependents from @cheukt 's new `updateWeakDependents` call in `completeConfig`) occurs when the "last round" is GTE the clock value or when no resources have weak dependencies.

> I think the "last round" is still at 0 and the "clock" is at 2 now?

Yep.

> So this would fail because we haven't initialized weak1 yet? ("There are no components with weak dependencies?")

Yep.

Reply:
> So this would fail because we haven't initialized weak1 yet? ("There are no components with weak dependencies?")

No, this failed because there are no dependencies on weak dependents in level 2, e.g. nothing depends on weak1